testsuite/testsuite_test.go

```diff
@@ -6,15 +6,99 @@ import (
 	"encoding/json"
 	"errors"
 	"fmt"
+	"io"
 	"math/rand"
 	"net/http"
 	"strings"
 	"testing"
 	"time"
 
 	"gotest.tools/v3/assert"
+
+	"github.com/ethereum/go-ethereum/crypto"
 )
 
+func TestTransactionAndCQLAndRead(t *testing.T) {
+
+	//Test persona
+	privateKey, err := crypto.GenerateKey()
+	assert.NilError(t, err)
+	signerAddr := crypto.PubkeyToAddress(privateKey.PublicKey).Hex()
+	username, deviceID, personaTag := triple(randomString())
+	c := newClient(t)
+	assert.NilError(t, c.registerDevice(username, deviceID))
+
+	resp, err := c.rpc("nakama/claim-persona", map[string]any{
+		"persona_tag":    personaTag,
+		"signer_address": signerAddr,
+	})
+	assert.NilError(t, err, "claim-persona failed")
+	assert.Equal(t, 200, resp.StatusCode, copyBody(resp))
+
+	assert.NilError(t, waitForAcceptedPersonaTag(c))
+	type CreatePlayerTxMsg struct{ Nickname string }
+	payload := CreatePlayerTxMsg{"Bob"}
+	resp, err = c.rpc("tx/game/create-player", payload)
+	assert.NilError(t, err)
+	body := copyBody(resp)
+	assert.Equal(t, 200, resp.StatusCode, body)
+
+	//Test CQL
+	type Data struct {
+		Nickname string `json:"nickname,omitempty"`
+		HP       int    `json:"HP,omitempty"`
+	}
+
+	type Item struct {
+		ID   int    `json:"id"`
+		Data []Data `json:"data"`
+	}
+	finalResults := []Item{}
+
+	for len(finalResults) <= 0 {
```

Comment:
What's the point of this for loop? I think you just want to do the single .rpc query and then assert that the unmarshaled results (in finalResults) is not empty.

Reply:
Because the test and server moves faster then the game loop. The game loop is impeded by an interval while the server and test move as fast as possible. Another way to do this is to use a delay, but changing the delay could make things inconsistent with the test.

Reply:
You should 1) add a comment documenting this and 2) Add a mechanism to time out after a sensible amount of time. https://pkg.go.dev/time#After could work, or just checking against a timestamp that you captured outside of the for loop.